### ISSUE TITLE
body limit exceeded error

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods_client_http_api#189]
### iRODS Version, OS and Version
4.3.1, almalinux8

### What did you try to do?
I tired to write 3MB file.

### Expected behavior
Data is written in irods. Or I expected to see buffer size error for writing.

### Observed behavior (including steps to reproduce, if applicable)

`http-api stdout | [2023-12-22 17:27:05.704] [P:681203] [error] [T:681203] on_read: Request constraint error: body limit exceeded
`

What my configuration has:
```
[root@dev-u0137480 /]# cat /etc/irods/http-api.config

    {
    "http_server": {
        "host": "0.0.0.0",
        "port": 9000,

        "log_level": "debug",

        "authentication": {
            "eviction_check_interval_in_seconds": 60,

            "basic": {
                "timeout_in_seconds": 3600
            }
        },

        "requests": {
            "threads": 3,
            "max_rbuffer_size_in_bytes": 8388608,
            "timeout_in_seconds": 30
        },

        "background_io": {
            "threads": 6
        }
    },

    "irods_client": {
        "host": "dev-u0137480.irods.t.icts.kuleuven.be",
        "port": 1247,
        "zone": "u0137480",

        "tls": {
            "client_server_policy": "CS_NEG_REQUIRE",
            "ca_certificate_file": "/etc/ssl/certs/ca-bundle.crt",
            "certificate_chain_file": "/etc/irods/ssl/server.crt",
            "dh_params_file": "/etc/irods/ssl/dhparams.pem",
            "verify_server": "cert"
        },

        "enable_4_2_compatibility": false,

        "proxy_admin_account": {
            "username": "rods",
            "password": "rods"
        },

        "connection_pool": {
            "size": 6,
            "refresh_timeout_in_seconds": 600,
            "max_retrievals_before_refresh": 16,
            "refresh_when_resource_changes_detected": true
        },

        "max_number_of_parallel_write_streams": 3,

        "max_number_of_bytes_per_read_operation": 8192,
        "buffer_size_in_bytes_for_write_operations": 8192,

        "max_number_of_rows_per_catalog_query": 15
    }
}
```
What is meant by body limit and by max_rbuffer_size_in_bytes?

**Comment by trel** [comment]
I think you'd have to increase your `buffer_size_in_bytes_for_write_operations` to take more than 8k in a single write.

I believe this is related to... https://github.com/irods/irods_client_http_api/issues/184

**Comment by mstfdkmn** [comment]
I forgot to add few things:
- I can write more than 8k in a single write. With the Go-http-client/1.1 
- I could write 2MB with my existing configuration values.
- As far as I test, the edge is somewhere between 2.5-3MB. 

my debug logs:
```
http-api stdout | [2023-12-22 19:21:29.395] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[2000000], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.395] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1991808], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.396] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1983616], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.396] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1975424], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.396] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1967232], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.397] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1959040], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.397] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1950848], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.397] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1942656], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.398] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1934464], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.398] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1926272], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.398] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1918080], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.399] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1909888], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.399] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1901696], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.399] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1893504], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.400] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1885312], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.400] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1877120], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.400] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1868928], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.401] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1860736], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.401] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1852544], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.401] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1844352], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.402] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1836160], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.402] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1827968], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.402] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1819776], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.403] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1811584], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.403] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1803392], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.404] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1795200], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.404] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1787008], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.404] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1778816], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.405] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1770624], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.405] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1762432], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.405] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1754240], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.406] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1746048], sending=[8192].

...

http-api stdout | [2023-12-22 19:21:29.476] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[58496], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.476] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[50304], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.476] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[42112], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.477] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[33920], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.477] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[25728], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.478] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[17536], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.478] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[9344], sending=[8192].
http-api stdout | [2023-12-22 19:21:29.478] [P:681203] [debug] [T:681251] op_write: Write buffer: remaining=[1152], sending=[1152].
```

**Comment by korydraughn** [comment]
`max_rbuffer_size_in_bytes` defines the max size of the request body for any incoming HTTP request. The config value is passed to `body_limit()`. See https://www.boost.org/doc/libs/1_81_0/libs/beast/doc/html/beast/ref/boost__beast__http__parser/body_limit.html for exact details. Your current config is set to about 8MB. I would expect uploading 3MB to work without any issues.

`buffer_size_in_bytes_for_write_operations` is the size of the internal buffer used for writing to a data object. This affects the number of calls made from the HTTP API to the iRODS server. Larger values result in fewer calls to the iRODS server at the expense of higher memory usage. Of course, smaller values result in more calls to the iRODS server, but less memory is used. This option is not the issue.

Can you post the code you ran?

**Comment by mstfdkmn** [comment]
Learning by testing - so this is also part of the effort to learn the lang! All errors due to the client remain mine. 
```
WriteDataObject("/u0137480/home/rods/test.img", "test.img")

func WriteDataObject(irodsPath, localPath string) {

    data := url.Values{}
    data.Set("op", "write")
    data.Set("lpath", irodsPath)
    contents, err := os.ReadFile(localPath)
    if err != nil {
        fmt.Println("File reading error", err)
        return
    }
    data.Set("count", fmt.Sprintf("%v", len(contents)))
    data.Set("bytes", string(contents))
    client := &http.Client{
        Timeout: time.Duration(30 * time.Second),
    }
    req, err := http.NewRequest(http.MethodPost, URL+"/data-objects", strings.NewReader(data.Encode()))
    if err != nil {
        fmt.Println("Error creating request:", err)
    }
    req.Header.Add("Authorization", "Bearer "+getToken())
    resp, err := client.Do(req)
    if err != nil {
        fmt.Println("Error sending request:", err)
    }
    defer resp.Body.Close()
    fmt.Println(resp.StatusCode)
}
```

Local file:
```
[irods@dev-u0137480 go-http-client]$ truncate -s 3MB test.img
[irods@dev-u0137480 go-http-client]$ ls -alh test.img
-rw-rw-r-- 1 irods irods 2.9M Dec 22 22:34 test.img
```

**Comment by mstfdkmn** [comment]
Also an example for curl. Specifying that size/bytes as a for count get stuck:

```
[irods@dev-u0137480 go-http-client]$ time curl -H "Authorization: Bearer $token" 'http://localhost:9000/irods-http-api/0.1.0/data-objects' -F  'op=write' -F 'lpath=/u0137480/home/rods/myfile.txt' -F 'count=8388608' -F 'bytes=abcdfg'
^C

real    1m38.440s
user    0m0.004s
sys     0m0.006s

http-api stdout | [2023-12-22 22:37:55.389] [P:722240] [info] [T:722286] op_write: client_info.username = [rods]
http-api stdout | [2023-12-22 22:37:55.411] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8388608], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.412] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8380416], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.412] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8372224], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.413] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8364032], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.413] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8355840], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.413] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8347648], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.414] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8339456], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.414] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8331264], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.414] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8323072], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.415] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8314880], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.415] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8306688], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.416] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8298496], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.416] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8290304], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.416] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8282112], sending=[8192].
http-api stdout | [2023-12-22 22:37:55.417] [P:722240] [debug] [T:722286] op_write: Write buffer: remaining=[8273920], sending=[8192].
```

**Comment by korydraughn** [comment]
I'm guessing the Go client is behaving? What happens if you bump the size of the local file to 7MB?

As for the curl example, the `count` value does not match the length of the `bytes` value. This causes the HTTP API to read pass what's actually available. That is certainly a bug we need to fix.

Regardless, the limit error message is a constraint that exists to protect against potential DDoS attacks, memory exhaustion, etc.

**Comment by korydraughn** [comment]
@mstfdkmn Any updates?

**Comment by korydraughn** [comment]
Resolved by #194.

By removing the need for the `count` parameter, there's no way for the server to read pass the end of the byte sequence. The length of the byte sequence is now recognized as the `count` internally.

Below is an example showing the `count` parameter is ignored.
```
$ token=$(curl -sX POST -u kory:rods 'http://localhost:9000/irods-http-api/0.1.0/authenticate')
$ curl 'http://localhost:9000/irods-http-api/0.1.0/data-objects' \
    -H "Authorization: Bearer $token" \
    -F 'op=write' \
    -F 'lpath=/tempZone/home/kory/httpfile' \
    -F 'count=5000' -F 'bytes=works' -s | jq
{
  "irods_response": {
    "status_code": 0
  }
}
$ istream read httpfile
works
```
Here's the logs confirming the results. Notice the `count` parameter's value is ignored _(see last line)_.
```
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Header: (Host, localhost:9000)
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Header: (User-Agent, curl/7.68.0)
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Header: (Accept, */*)
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Header: (Authorization, Bearer 891b7fd3-ea34-4ef0-a2bf-955d9c274ba7)
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Header: (Content-Length, 461)
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Header: (Content-Type, multipart/form-data; boundary=------------------------d7c9b60281626931)
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Method: POST
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Version: 11
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Target: /irods-http-api/0.1.0/data-objects
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Keep Alive: true
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Has Content Length: true
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Chunked: false
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] on_read: Needs EOF: false
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] PARAM = (name=boundary, value=------------------------d7c9b60281626931)
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] PARAM = (name=name, value=op)
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] END OF HEADERS
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] parse_multipart_form_data: CAPTURED CONTENT SIZE = [5]
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] END OF CONTENT
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] PARAM = (name=name, value=lpath)
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] END OF HEADERS
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] parse_multipart_form_data: CAPTURED CONTENT SIZE = [28]
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] END OF CONTENT
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] PARAM = (name=name, value=count)
[2024-01-10 15:24:35.720] [P:3901514] [trace] [T:3901602] END OF HEADERS
[2024-01-10 15:24:35.720] [P:3901514] [debug] [T:3901602] parse_multipart_form_data: CAPTURED CONTENT SIZE = [4]
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901602] END OF CONTENT
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901602] PARAM = (name=name, value=bytes)
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901602] END OF HEADERS
[2024-01-10 15:24:35.721] [P:3901514] [debug] [T:3901602] parse_multipart_form_data: CAPTURED CONTENT SIZE = [5]
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901602] END OF CONTENT
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901602] END OF REQUEST
[2024-01-10 15:24:35.721] [P:3901514] [debug] [T:3901602] resolve_client_identity: Authorization value: [Bearer 891b7fd3-ea34-4ef0-a2bf-955d9c274ba7]
[2024-01-10 15:24:35.721] [P:3901514] [debug] [T:3901602] resolve_client_identity: Bearer token: [891b7fd3-ea34-4ef0-a2bf-955d9c274ba7]
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901602] resolve_client_identity: Client is authenticated.
[2024-01-10 15:24:35.721] [P:3901514] [info] [T:3901599] op_write: client_info.username = [kory]
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901599] op_write: Opening data object [/tempZone/home/kory/httpfile] for write.
[2024-01-10 15:24:35.721] [P:3901514] [trace] [T:3901599] op_write: (write) Initializing for single buffer write.
[2024-01-10 15:24:35.725] [P:3901514] [trace] [T:3901599] get_connection: Changing identity associated with connection to [kory].
[2024-01-10 15:24:35.725] [P:3901514] [trace] [T:3901599] get_connection: Successfully changed identity associated with connection to [kory].
[2024-01-10 15:24:35.790] [P:3901514] [debug] [T:3901599] op_write: Write buffer: remaining=[5], sending=[5].
```
